frontend/javascripts/oxalis/view/right-border-tabs/connectome_tab/connectome_view.js

```diff
@@ -0,0 +1,784 @@
+      yield* mapAndFilterTreeData(node.children, callback, condition);
+    }
+  }
+}
+
+function ensureTypeToString(synapseTypesAndNames) {
+  const { synapseTypes, typeToString } = synapseTypesAndNames;
+  if (typeToString.length === 0) {
+    Toast.error(`Couldn't read synapseTypes mapping. Please add a json file containing the synapseTypes next to the connectome file.
+The format should be: \`{
+  "synapse_type_names": [ "type1", "type2", ... ]
+}\``);
+    // Create mocked synapse type strings
+    const largestSynapseTypeId = Math.max(...synapseTypes);
+    const mockedTypeToString = [...Array(largestSynapseTypeId + 1).keys()].map(i => `type${i + 1}`);
+    return {
+      synapseTypes,
+      typeToString: mockedTypeToString,
+    };
+  }
+  return synapseTypesAndNames;
+}
+
+class ConnectomeView extends React.Component<Props, State> {
+  skeletonId: ?number;
+  state = {
+    connectomeData: null,
+    filteredConnectomeData: null,
+    availableSynapseTypes: [],
+    checkedKeys: [],
+    expandedKeys: [],
+  };
+
+  componentDidMount() {
+    this.initializeSkeleton();
+    this.fetchConnections();
+  }
+
+  componentDidUpdate(prevProps: Props, prevState: State) {
+    if (
+      prevProps.activeAgglomerateIds !== this.props.activeAgglomerateIds ||
+      prevProps.currentConnectomeFile !== this.props.currentConnectomeFile
+    ) {
+      this.fetchConnections();
+    }
+    if (prevProps.segmentationLayer !== this.props.segmentationLayer) {
+      this.maybeUpdateSkeleton(prevProps.segmentationLayer);
+    }
+    if (prevState.filteredConnectomeData !== this.state.filteredConnectomeData) {
+      this.updateSynapseTrees(prevState.filteredConnectomeData);
+    }
+    if (
+      prevState.connectomeData !== this.state.connectomeData ||
+      prevState.filteredConnectomeData !== this.state.filteredConnectomeData ||
+      prevState.checkedKeys !== this.state.checkedKeys
+    ) {
+      this.updateAgglomerateTrees(
+        prevState.connectomeData,
+        prevState.filteredConnectomeData,
+        prevState.checkedKeys,
+      );
+    }
+  }
+
+  componentWillUnmount() {
+    const { segmentationLayer } = this.props;
+    if (segmentationLayer == null) return;
+
+    this.removeSkeleton(segmentationLayer);
+    this.reset();
+  }
+
+  reset(segmentationLayer?: ?APISegmentationLayer) {
+    if (segmentationLayer == null) {
+      segmentationLayer = this.props.segmentationLayer;
+    }
+    if (segmentationLayer != null) {
+      Store.dispatch(setActiveConnectomeAgglomerateIdsAction(segmentationLayer.name, []));
+    }
+
+    this.setState({
+      connectomeData: null,
+      filteredConnectomeData: null,
+      checkedKeys: [],
+      expandedKeys: [],
+    });
+  }
+
+  initializeSkeleton() {
+    const { segmentationLayer } = this.props;
+    if (segmentationLayer == null) return;
+
+    Store.dispatch(initializeConnectomeTracingAction(segmentationLayer.name));
+    this.skeletonId = getSceneController().addSkeleton(
```

Comment:
Maybe assert that `this.skeletonId == null` to ensure that `removeSkeleton` was called before?